PLT-7925 Add a first version of the marconi key concepts documentation (#218)
input-output-hk/marconi: doc/read-the-docs-site/architecture/key-concepts.rst

Marconi Key concepts
====================

We introduce here the key concepts of Marconi, how they are related and how they
map to the architecture

Indexer
-------

The key component of Marconi is an ``indexer``.
An indexer can be any type that implement the ``IsIndex`` typeclass. The

> definition of ``IsIndex`` is (very close) to this implementation:

Author: Yes, I'm removing the implementation word, it's a simplified definition. Rewording it.

.. code-block:: haskell
   class (Monad m) => IsIndex m event indexer where

     index
       :: (Eq (Point event))
       => Timed (Point event) (Maybe event)
       -> indexer event
       -> m (indexer event)

    // …

    rollback :: (Ord (Point event)) => Point event -> indexer event -> m (indexer event)

Let's discuss the type variables first:

- ``m`` is the monadic context of the indexer.
- ``event`` defines the input type of the indexer, we usually stores these
  events, but an indexer can also decide to preprocess them and store a

>   different type.

Author: It can be done without preprocessing, by putting logic into the `index` function. Alhough, I understand that using _preprocess_ is confusing. Changing it to process.

- ``indexer`` maintain the indexer state and provide the necessary information
  to perform the indexing operations.

> The two keys operation that we have to deal with are ``index`` and ``rollback``.

Reviewer: Typo?

```suggestion
The two key operations that we have to deal with are ``index`` and ``rollback``.
```

In a blockchain context, we call ``index`` when a new block comes in, to index
the event that correspond to that block. ``rollback`` is called when we spot a
rollback: some of the indexed blocks aren't aligned with the blockchain
consensus and we need to unindex them.

``index`` takes a ``Timed (Point event) (Maybe event)`` and the indexer that
provides the indexing context. ``Point event`` is a type family that defines how
we identify the block that produces the `event`.
In the context of Cardano, a ``Point event`` will almost always be a
``ChainPoint``: a blockHeaderHash to identify a block and the slot at which the
block was issued.
The ``event`` is wrapped in a ``Maybe`` to express the possibility that a block
may not contain any relevant information for a given indexer. It still important

> to index a block with now relevant event to keep track of the progress of an

Author: It's _no relevant_, sorry for this. :)

indexer, we'll come back on this later.

``rollback`` takes a point in time. This is the point to which we must rollback
the indexer. It mens that we must put the indexer back to the state it was when
we indexed the block at the given point.

``IsIndex`` provides other functions. ``indexAll`` and ``indexAllDescending``
that allows indexing of a list of events.
A default imprementation is provided for these functions, and they can be
overriden to provide a more efficient implementation.

It also provides a ``setLastStablePoint`` method that allows the indexer to keep
track of the last stable point progress and if needed, to react accordingly.

Coordinator and Workers
-----------------------

> A common indexing scenario is to run several indexers in parrallel to index

Reviewer 2: parrallel -> parallel

to index -> in order to index

> different parts of a same event, ideally we want to organise our indexers a

Reviewer: Typo?
```suggestion
different parts of a same event, ideally we want to organise our indexers as
```

a tree, each node doing a part of the processing before propagating the

> resulting events, and the leaves being indexers that stores the relevant part of

Reviewer: ```suggestion
resulting events, and the leaves being indexers that store the relevant part of
```

these events.

> In Marconi, the indexers can be organised thanks to a ``Coordinator``.

Reviewer 2: organised -> grouped?

> A coordinator is a special type of indexer that propagate incoming actions

Reviewer: ```suggestion
A coordinator is a special type of indexer that propagates incoming actions
```

> (index, rollback) to a list of indexers, and handle the lifecycle of these

Reviewer: ```suggestion
(index, rollback) to a list of indexers, and handles the lifecycle of these
```

The type of a worker (slightly simplified) is the following:

.. code-block:: haskell
   data Worker input point = forall indexer event n.
     ( IsIndex n event indexer
     , Point event ~ point
     ) =>
     Worker
     { workerName :: Text
     -- ^ use to identify the worker in logs
     , workerState :: MVar (indexer event)
     -- ^ the indexer controlled by this worker
     , transformInput :: Preprocessor (ExceptT IndexerError m) point input event
     -- ^ adapt the input event givent by the coordinator to the worker type
     , hoistError :: forall a. n a -> ExceptT IndexerError IO a
     -- ^ adapt the monadic stack of the indexer to the one of the worker
     }

You don't need to understand this type declaration in details. Here are the
important bits:

- We hide the indexer representation under the worker type. Any concrete indexer
  type will work as long as it implements the ``IsIndex`` typeclass.
- The indexer is put in an ``MVar`` to allow access to it from other threads.
  We'll come back to it later.
- We provide a preprocessor, which can be seen as a functon that transforms
  the input send by the coordinator into events.

> - ``hoistError`` ensures that we now how to translate the base monad of the

Reviewer: ```suggestion
- ``hoistError`` ensures that we know how to translate the base monad of the
```

  indexer into ``ExceptT IndexerError IO`` which is the base monad of a worker.

Once we have a list of workers, we can use it to create a coordinator using the
``mkCoordinator`` function.

> This function inialise each worker, creating a dedicated thread for each worker

Reviewer 2: initialises

where the worker will wait for incoming action to perform and notify the
coordinator when the action is performed.

The coordinator monitors each of the threads and, if one of the worker encounter
an error, it will try to close all the other workers nicely.

``Coordinator`` itself implements ``IsIndex`` and thus we can itself be wrapped
in a worker.
Thanks to it, we can create a whole hierarchy of indexers that can control from
a main coordinator.

Preprocessor
------------

> We saw in the coordinator and workers section that workers takes a preprocessor.

Reviewer: ```suggestion
We saw in the coordinator and workers section that workers take a preprocessor.
```

As stated in this section, the preprocessor type can be viewed as a stateful
function that transform the action sent to an indexer.
It's type is isomorphic to:

.. code-block:: haskell
   StateT s m ([ProcessedInput point a] -> [ProcessedInput point b])

> It's the first time we encounter ``ProcessedInput`` so it worths going through

Reviewer: ```suggestion
It's the first time we encounter ``ProcessedInput`` so it is worth going through
```

its definition:

.. code-block:: haskell
   data ProcessedInput point event
     = Rollback point
     | Index (Timed point (Maybe event))
     | IndexAllDescending (NonEmpty (Timed point (Maybe event)))
     | StableAt point
     | Stop

It is mostly a reified version (functions expressed as data) of most of the

> ``IsIndex`` functions plus a ``Stop`` construct that allow us to stop a worker.

Reviewer: ```suggestion
``IsIndex`` functions plus a ``Stop`` construct that allows us to stop a worker.
```

So the goal of a preprocessor is to take a list of actions that must be sent to
a worker and to transform this list.
It can be either to filter out some actions or to add some actions to the list,
based on the internal state ``s`` of the preprocessor.

Transformers
------------

Indexer transformers (or, shorter, transformers) are used to alter the behaviour
of an existing indexer.
The name comes from the monad transformers concept, and it was chosen because
monad and indexer transformers have similarities, as both aims at adding
capabilities to a base implementation.

> A transformer carry a state and an underlying indexer. It usually implements

Reviewer: ```suggestion
A transformer carries a state and an underlying indexer. It usually implements
```

the same typeclass than the one of the indexer it is supposed to wrap and modify

> some of their behaviour.

Reviewer: Can you reword this please? It's not clear to me what this means.

Preprocessor vs Transformers
----------------------------

There's a lot of similarities between preprocessor and transformers.
There are two major differences:

- ``Preprocessor`` state is not exposed and can't be accessed from the outside
  while ``Transformers`` can expose their state.
  It makes preprocessors slightly less powerful.
- ``Preprocessor`` doesn't rely on typeclasses implementation. As a consequence,
  they are easier to write and easier to compose in the general case.
  Furthermore, they don't change the type of the indexers.

Tracking progress
-----------------

In Marconi, each indexer is independent.
You can reuse data from another indexer in your application, add or remove
indexers from one run to another, without compromising the other indexers.
A consequence is that each indexer must track its synchronisation progress.

> When there's a call on ``index`` the indexer keep track of its last

Reviewer: ```suggestion
When there's a call on ``index`` the indexer keeps track of its last
```

synchronisation point
When there's a call to ``setLastStablePoint``, it is supposed to keep track
of the given point as well.

In many scenarios, it is useful to have access to the last sync point and last
stable point of an indexer.

> If you need to query different indexer for example, you may want to query them

Reviewer: ```suggestion
If you need to query different indexers for example, you may want to query them
```

> at the same point in time, to ensure that you get consistent result.

Reviewer: ```suggestion
at the same point in time, to ensure that you get consistent results.
```

When you restart your indexers, you may want to access their last stable point
to know at which blocks you must restart your synchronisation.

> To expose these information, we usually implements the ``IsSync`` typeclass.

Reviewer: ```suggestion
To expose this information, we usually implement the ``IsSync`` typeclass.
```

Implementing this typeclass is required to put an indexer in a worker.

Queryable
---------

The whole point of an indexer is to expose information about the event
they index.
In Marconi, it's done through the ``Queryable`` typeclass, which has the
following deifinition:

.. code-block:: haskell
   class Queryable m event query indexer where
     query
       :: (Ord (Point event))
       => Point event
       -> query
       -> indexer event
       -> m (Result query)

To define a queryable instance, you need to provide the context ``m`` in
which it operates, the ``event`` that the indexer must handle to be able to
answer the query, the query type and the indexer implementation that can answer
the query.

Then, you need to implement the ``query`` method, which takes a point, a query
and an indexer to provide a ``Result query``. ``Result`` is a type family that
associate a query to its result.
The ``point`` needed by query define both a point that the indexer must have
reached and the upper bound of the result we consider, if applicable.
When you need to do several queries to different indexers, passing the same

> point to the different queries ensure you that the results are consistent.

Reviewer: ```suggestion
point to the different queries ensures that the results are consistent.
```

Once we have a list of workers, we can use it to create a coordinator using the
``mkCoordinator`` function.
This function inialise each worker, creating a dedicated thread for each worker
where the worker will wait for incoming action to perform and notify the
coordinator when the action is performed.

The coordinator monitors each of the threads and, if one of the worker encounter
an error, it will try to close all the other workers nicely.

``Coordinator`` itself implements ``IsIndex`` and thus we can itself be wrapped
in a worker.
Thanks to it, we can create a whole hierarchy of indexers that can control from
a main coordinator.

Preprocessor
------------

We saw in the coordinator and workers section that workers take a preprocessor.
As stated in this section, the preprocessor type can be viewed as a stateful
function that transform the action sent to an indexer.
It's type is isomorphic to:

.. code-block:: haskell

   StateT s m ([ProcessedInput point a] -> [ProcessedInput point b])

It's the first time we encounter ``ProcessedInput`` so it is worth going through
its definition:

.. code-block:: haskell

   data ProcessedInput point event
     = Rollback point

>      | Index (Timed point (Maybe event))

Reviewer 2: to transform the event?

     | IndexAllDescending (NonEmpty (Timed point (Maybe event)))

>      | StableAt point

Reviewer 2: Add a example of a preprocessor? Like the resume?

     | Stop

It is mostly a reified version (functions expressed as data) of most of the
``IsIndex`` functions plus a ``Stop`` construct that allows us to stop a worker.
So the goal of a preprocessor is to take a list of actions that must be sent to
a worker and to transform this list.
It can be either to filter out some actions or to add some actions to the list,
based on the internal state ``s`` of the preprocessor.

Transformers
------------

Indexer transformers (or, shorter, transformers) are used to alter the behaviour

> of an existing indexer.

Reviewer 2: Again, maybe an example of a transformer? Maybe the Catchup?

The name comes from the monad transformers concept, and it was chosen because
monad and indexer transformers have similarities, as both aims at adding
capabilities to a base implementation.

A transformer carries a state and an underlying indexer.
The typeclass instances of the transformer can then add extra logic to this

> underlying indexer.

Reviewer 2: What do you mean by "the outside"? An example here would be helpful.

Author: I changed it before you submitted your review so I can't remember. An example can still be useful though, will think about it.

Author: The problem is that a minimal example is already a lot of code: We need the datatype, the instance it modifies and (ideally) a bunch of derived instances (we can skip them but the example would miss an important part).
If the latest version isn't clear enough, we can either add an extra page about it later or refer to an example in the codebase?

Preprocessor vs Transformers
----------------------------

> There's a lot of similarities between preprocessor and transformers.

Reviewer 2: Can some preprocessors be rewritten as transformers? Or vis-versa? Maybe mention when that would be the case.

Author: Yes, added.

There are two major differences:

> 

Reviewer 2: ```suggestion
Tracking synchronisation progress
```

You can reuse data from another indexer in your application, add or remove
indexers from one run to another, without compromising the other indexers.
A consequence is that each indexer must track its synchronisation progress.
When there's a call on ``index`` the indexer keeps track of its last
synchronisation point
When there's a call to ``setLastStablePoint``, it is supposed to keep track
of the given point as well.

In many scenarios, it is useful to have access to the last sync point and last
stable point of an indexer.
If you need to query different indexers for example, you may want to query them
at the same point in time, to ensure that you get consistent results.
When you restart your indexers, you may want to access their last stable point
to know at which blocks you must restart your synchronisation.

To expose this information, we usually implement the ``IsSync`` typeclass.
Implementing this typeclass is required to put an indexer in a worker.

Queryable
---------

The whole point of an indexer is to expose information about the event
they index.
In Marconi, it's done through the ``Queryable`` typeclass, which has the
following deifinition:

.. code-block:: haskell

> 

Reviewer 2: ```suggestion
To define a ``Queryable`` instance, you need to provide the context ``m`` in
```

   class Queryable m event query indexer where
     query
       :: (Ord (Point event))
       => Point event
       -> query
       -> indexer event
       -> m (Result query)

> 

Reviewer 2: ```suggestion
The ``point`` needed by the query defines both a point that the indexer must have
```

Marconi Key concepts
====================

We introduce here the key concepts of Marconi, how they are related and how they
map to the architecture.

Indexer
-------

The key component of Marconi is an ``indexer``.
An indexer can be any type that implements the ``IsIndex`` typeclass.
The definition of ``IsIndex`` is (very close to) this one:

.. code-block:: haskell

   class (Monad m) => IsIndex m event indexer where

     index
       :: (Eq (Point event))
       => Timed (Point event) (Maybe event)
       -> indexer event
       -> m (indexer event)

    -- …

    rollback :: (Ord (Point event)) => Point event -> indexer event -> m (indexer event)

Let's discuss the type variables first:

- ``m`` is the monadic context of the indexer.
- ``event`` defines the input type of the indexer, we usually store these
  events, but an indexer can also decide to process them and store a
  different type.
- ``indexer`` maintain the indexer state and provide the necessary information
  to perform the indexing operations.

> The two keys operations that we have to deal with are ``index`` and ``rollback``.

Reviewer 2: keys -> key

In a blockchain context, we call ``index`` when a new block comes in, to index

> the event that correspond to that block. ``rollback`` is called when we spot a

Reviewer 2: correspond -> corresponds

rollback: some of the indexed blocks aren't aligned with the blockchain
consensus and we need to unindex them.

``index`` takes a ``Timed (Point event) (Maybe event)`` and the indexer that
provides the indexing context. ``Point event`` is a type family that defines how

> we identify the block that produces the `event`.

Reviewer 2: Maybe change to "we identify the point of a block that produces the event"?

In the context of Cardano, a ``Point event`` will almost always be a
``ChainPoint``: a blockHeaderHash to identify a block and the slot at which the
block was issued.
The ``event`` is wrapped in a ``Maybe`` to express the possibility that a block
may not contain any relevant information for a given indexer. It still important
to index a block with no relevant event to keep track of the progress of an

> indexer, we'll come back on this later.

Reviewer 2: , we'll -> , which we'll

``rollback`` takes a point in time. This is the point to which we must rollback
the indexer. It means that we must put the indexer back to the state it was when
we indexed the block at the given point.

``IsIndex`` provides other functions. ``indexAll`` and ``indexAllDescending``
that allows indexing of a list of events.
A default imprementation is provided for these functions, and they can be
overriden to provide a more efficient implementation.

It also provides a ``setLastStablePoint`` method that allows the indexer to keep
track of the last stable point progress and if needed, to react accordingly.

Coordinator and Workers
-----------------------

A common indexing scenario is to run several indexers in parallel to index

> different parts of a same event, ideally we want to organise our indexers as

Reviewer 2: event, ideally -> event. Ideally,

     -- ^ adapt the input event givent by the coordinator to the worker type
     , hoistError :: forall a. n a -> ExceptT IndexerError IO a
     -- ^ adapt the monadic stack of the indexer to the one of the worker
     }

You don't need to understand this type declaration in details. Here are the
important bits:

- We hide the indexer representation under the worker type. Any concrete indexer
  type will work as long as it implements the ``IsIndex`` typeclass.
- The indexer is put in an ``MVar`` to allow access to it from other threads.
  We'll come back to it later.
- We provide a preprocessor, which can be seen as a functon that transforms
  the input send by the coordinator into events.
- ``hoistError`` ensures that we know how to translate the base monad of the
  indexer into ``ExceptT IndexerError IO`` which is the base monad of a worker.

Once we have a list of workers, we can use it to create a coordinator using the
``mkCoordinator`` function.
This function inialise each worker, creating a dedicated thread for each worker
where the worker will wait for incoming action to perform and notify the
coordinator when the action is performed.

> The coordinator monitors each of the threads and, if one of the worker encounter

Reviewer 2: encounter -> encounters

As stated in this section, the preprocessor type can be viewed as a stateful
function that transform the action sent to an indexer.
It's type is isomorphic to:

.. code-block:: haskell

   StateT s m ([ProcessedInput point a] -> [ProcessedInput point b])

It's the first time we encounter ``ProcessedInput`` so it is worth going through
its definition:

.. code-block:: haskell

   data ProcessedInput point event
     = Rollback point
     | Index (Timed point (Maybe event))
     | IndexAllDescending (NonEmpty (Timed point (Maybe event)))
     | StableAt point
     | Stop

It is mostly a reified version (functions expressed as data) of most of the
``IsIndex`` functions plus a ``Stop`` construct that allows us to stop a worker.
So the goal of a preprocessor is to take a list of actions that must be sent to
a worker and to transform this list.
It can be either to filter out some actions or to add some actions to the list,
based on the internal state ``s`` of the preprocessor.

Transformers
------------

Indexer transformers (or, shorter, transformers) are used to alter the behaviour
of an existing indexer.
The name comes from the monad transformers concept, and it was chosen because

> monad and indexer transformers have similarities, as both aims at adding

Reviewer 2: aims -> aim

underlying indexer.

Preprocessor vs Transformers
----------------------------

There's a lot of similarities between preprocessor and transformers.
There are two major differences:

- ``Preprocessor`` state is not exposed and can't be accessed from the outside
  while ``Transformers`` can expose their state.
  It makes preprocessors slightly less powerful.
- ``Preprocessor`` doesn't rely on typeclasses implementation. As a consequence,
  they are easier to write and easier to compose in the general case.
  Furthermore, they don't change the type of the indexers.

Despite these differences, preprocessors and transformers are closely related
and in most scenarios, you should be able to rewrite a preprocessor as a
transformer or the opposite.

Tracking synchronisation progress
---------------------------------

In Marconi, each indexer is independent.
You can reuse data from another indexer in your application, add or remove
indexers from one run to another, without compromising the other indexers.
A consequence is that each indexer must track its synchronisation progress.
When there's a call on ``index`` the indexer keeps track of its last

> synchronisation point

Reviewer 2: point -> point.

When there's a call to ``setLastStablePoint``, it is supposed to keep track
of the given point as well.

In many scenarios, it is useful to have access to the last sync point and last
stable point of an indexer.
If you need to query different indexers for example, you may want to query them
at the same point in time, to ensure that you get consistent results.
When you restart your indexers, you may want to access their last stable point
to know at which blocks you must restart your synchronisation.

To expose this information, we usually implement the ``IsSync`` typeclass.

> Implementing this typeclass is required to put an indexer in a worker.

Reviewer 2: is required to put -> is required in order to put

Queryable
---------

The whole point of an indexer is to expose information about the event
they index.
In Marconi, it's done through the ``Queryable`` typeclass, which has the
following deifinition:

.. code-block:: haskell

   class Queryable m event query indexer where
     query
       :: (Ord (Point event))
       => Point event
       -> query
       -> indexer event
       -> m (Result query)

To define a ``Queryable`` instance, you need to provide the context ``m`` in
which it operates, the ``event`` that the indexer must handle to be able to
answer the query, the query type and the indexer implementation that can answer
the query.

Then, you need to implement the ``query`` method, which takes a point, a query
and an indexer to provide a ``Result query``. ``Result`` is a type family that

> associate a query to its result.

Reviewer 2: associate -> associates

> The ``point`` needed by query defines both a point that the indexer must have

Reviewer 2: query -> ``query``

reached and the upper bound of the result we consider, if applicable.
When you need to do several queries to different indexers, passing the same
point to the different queries ensures that the results are consistent.

In many situations, we just want access to the freshest information of an
indexer. In these scenarios, one can use the ``queryLatest`` function.

> ``queryLatest`` requires the indexer to implement both ``queryable`` and

Reviewer 2: queryable -> Queryable
